games/minimal/mods/experimental/commands.lua

```diff
@@ -0,0 +1,215 @@
+			"button_exit[2.5,7;2,1;button2;Exit Button]")
+		return true, "Done."
+	end,
+})
+
+minetest.register_chatcommand("test_bulk_set_node", {
+	params = "",
+	description = "Test: Bulk-set 9×9×9 stone nodes",
+	func = function(name, param)
+		local player = minetest.get_player_by_name(name)
+		if not player then
+			return false, "No player."
+		end
+		local pos_list = {}
+		local ppos = player:get_pos()
+		local i = 1
+		for x=2,10 do
+			for y=2,10 do
+				for z=2,10 do
+					pos_list[i] = {x=ppos.x + x,y = ppos.y + y,z = ppos.z + z}
+					i = i + 1
+				end
+			end
+		end
+		minetest.bulk_set_node(pos_list, {name = "mapgen_stone"})
+		return true, "Done."
+	end,
+})
+
+minetest.register_chatcommand("bench_bulk_set_node", {
+	params = "",
+	description = "Benchmark: Bulk-set 99×99×99 stone nodes",
+	func = function(name, param)
+		local player = minetest.get_player_by_name(name)
+		if not player then
+			return false, "No player."
+		end
+		local pos_list = {}
+		local ppos = player:get_pos()
+		local i = 1
+		for x=2,100 do
+			for y=2,100 do
+				for z=2,100 do
+					pos_list[i] = {x=ppos.x + x,y = ppos.y + y,z = ppos.z + z}
+					i = i + 1
+				end
+			end
+		end
+
+		minetest.chat_send_player(name, "Benchmarking minetest.bulk_set_node. Warming up ...");
+
+		-- warm up with stone to prevent having different callbacks
+		-- due to different node topology
+		minetest.bulk_set_node(pos_list, {name = "mapgen_stone"})
+
+		minetest.chat_send_player(name, "Warming up finished, now benchmarking ...");
+
+		local start_time = minetest.get_us_time()
+		for i=1,#pos_list do
+			minetest.set_node(pos_list[i], {name = "mapgen_stone"})
+		end
+		local middle_time = minetest.get_us_time()
+		minetest.bulk_set_node(pos_list, {name = "mapgen_stone"})
+		local end_time = minetest.get_us_time()
+		local msg = string.format("Benchmark results: minetest.set_node loop: %.2f ms; minetest.bulk_set_node: %.2f ms",
+			((middle_time - start_time)) / 1000,
+			((end_time - middle_time)) / 1000
+		)
+		return true, msg
+	end,
+})
+
+local function advance_pos(pos, start_pos, advance_z)
+	if advance_z then
+		pos.z = pos.z + 2
+		pos.x = start_pos.x
+	else
+		pos.x = pos.x + 2
+	end
+	if pos.x > 30900 or pos.x - start_pos.x > 46 then
+		pos.x = start_pos.x
+		pos.z = pos.z + 2
+	end
+	if pos.z > 30900 then
+		-- We ran out of space! Aborting
+		aborted = true
+		return false
+	end
+	return pos
+end
+
+local function place_nodes(param)
+	local nodes = param.nodes
+	local name = param.name
+	local pos = param.pos
+	local start_pos = param.start_pos
+	table.sort(nodes)
+	minetest.chat_send_player(name, "Placing nodes …")
+	local nodes_placed = 0
+	for n=1, #nodes do
+		local itemstring = nodes[n]
+		local def = minetest.registered_nodes[itemstring]
+		local p2_max = 0
+		if param.param ~= "no_param2" then
+			-- Also test the param2 values of the nodes
+			-- ... but we only use permissible param2 values
+			if def.paramtype2 == "wallmounted" then
```

Comment:
If-chain can be replaced by using lookup table (small code quality note).

Reply:
`drawtype` checks are needed too, which defeats lookup tables. The code works as-is and is acceptable in size. No change needed.